### PR TITLE
An attempt to clean up tesla code

### LINKED ISSUE #50506: Teslas blow up random shit in contaiment
Setting up a tesla in the delta containment will cause this. No idea what causes it, I assume it has to do with the singularity parent path. We shall see.

### PR DESCRIPTION
**Pull request by LemonInTheDark**
## About The Pull Request

Adds an override to tesla's consume() to prevent the tesla from calling consume() on what seems to be it's own energy balls, blowing up the containment.
Refactors the way tesla_zap deals with type, instead of dealing with type with a var for each type, we store the type integer, so we can clean up the zap_act part and something something lower the amount of vats we have to create on each zap.
Moves the base of zap_act to /atom so we don't need to type the default target var on most zaps.
Uses this in the same manner in supermatter_zap
*Should* make power output of tesla zaps more sane, mostly from tesla coils. I'll go into this more in future.

Now onto the more interesting changes.
We no longer track the closest object. We can do this because of how byond outputs objects with range()/view().
It outputs things in a spiral, starting from the center. [See](https://tgstation13.org/phpBB/viewtopic.php?f=5&t=25922)
Let me know if you want me to change this. It should speed up things slightly, it's mostly a cleanliness thing at this point.

And finally onto what I didn't change.
oview() for some insane reason cares about darkness. I've tried dview() and a modified version of that doview(), and neither worked. It didn't seem to return any objects from the for loop.  I plan on doing more testing on this in future.

## Why It's Good For The Game

Fixes #50506 
Cleans some stuff up, fixes some fucking weird shit. Documents some strange behavior.
Maybe now b**mer st##ion feature coders will stop being toxic towards my code.

## Changelog
:cl:
fix: Fixes teslas not containing properly sometimes
fix: Maybe makes tesla coil power output more sane, not totally sure myself
refactor: Refactored tesla code a bit
/:cl:

### CODE REVIEW

File: code/modules/power/tesla/energy_ball.dm [hunk 1]
```diff
@@ -211,123 +215,104 @@
...
 										/obj/structure/frame/machine))
 
-	for(var/A in typecache_filter_multi_list_exclusion(oview(source, zap_range+2), things_to_shock, blacklisted_tesla_types))
+	//Ok so we are making an assumption here. We assume that view() still calculates from the center out.
+	//This means that if we find an object we can assume it is the closest one of its type. This is somewhat of a speed increase.
+	//This also means we have no need to track distance, as the doview() proc does it all for us.
+
+	//Darkness fucks oview up hard. I've tried dview() but it doesn't seem to work
+	//I hate existance
+	for(var/atom/A in typecache_filter_multi_list_exclusion(oview(zap_range+2, source), things_to_shock, blacklisted_tesla_types))
```
Comment (spookydonut) [inline code/modules/power/tesla/energy_ball.dm:224]: make this typeless

Comment (spookydonut) [inline code/modules/power/tesla/energy_ball.dm:224]:
```suggestion
	for(var/i in typecache_filter_multi_list_exclusion(oview(zap_range+2, source), things_to_shock, blacklisted_tesla_types))
		var/atom/A = i
```

**Review by spookydonut — CHANGES_REQUESTED** [review]

File: code/modules/power/tesla/energy_ball.dm (hunk 1 above)
Comment (LemonInTheDark) [reply code/modules/power/tesla/energy_ball.dm:224]: Finally a micro-op I can get behind.

Comment (spookydonut) [reply code/modules/power/tesla/energy_ball.dm:224]: its not a micro opt its a normal opt

**Comment by Rohesie** [comment]
Are you planning to include turfs in `zap_act()`? Else just use `atom/movable`

**Comment by LemonInTheDark** [comment]
I'm gonna go with what spooky suggested. Using atom/movable would actually be a little bit slower because of how the list gets culled, and how typecache_filter_multi_list_exclusion works. At least I'm reasonably sure so.

File: code/modules/power/supermatter/supermatter.dm [hunk 2]
```diff
@@ -919,8 +919,8 @@ GLOBAL_DATUM(main_supermatter_engine, /obj/machinery/power/supermatter_crystal)
 	//If the strength of the zap decays past the cutoff, we stop
 	if(zap_str < zap_cutoff)
 		return
-	var/datum/target
-	var/target_type = (LOWEST)
+	var/atom/target
+	var/target_type = 0
```
Comment (Rohesie) [inline code/modules/power/supermatter/supermatter.dm:923]: Keep a define here instead of a number.

Comment (LemonInTheDark) [reply code/modules/power/supermatter/supermatter.dm]: sure

File: code/modules/power/tesla/energy_ball.dm (hunk 1 above)
Comment (LemonInTheDark) [reply code/modules/power/tesla/energy_ball.dm]: done

**Comment by Rohesie** [comment]
Alright, have you tested this locally to perform as you expect it to?

**Comment by LemonInTheDark** [comment]
I have. My main concern is the reliance on the order view() outputs objects.

Is that an issue? It's a thing in this version, but from what I've heard people say it hasn't always been.

**Comment by Rohesie** [comment]
Looks like `view()` lists things ordered by proximity. It scans one range-circle at a time. Which is an assumption this code does, correctly.
Not the best behavior to depend on, mind you, but it is what it is.

**Pull request merged**